src/transport/manager/peer_state.rs

```diff
@@ -0,0 +1,928 @@
+            } =>
+                if dial_record.connection_id == connection_id {
+                    *self = Self::Connected {
+                        record: record.clone(),
+                        secondary: None,
+                    };
+                    return true;
+                },
+
+            Self::Disconnected {
+                dial_record: Some(dial_record),
+            } =>
+                if dial_record.connection_id == connection_id {
+                    *self = Self::Disconnected { dial_record: None };
+                    return true;
+                },
+
+            _ => (),
+        };
+
+        false
+    }
+
+    /// Returns `true` if the connection should be accepted by the transport manager.
+    pub fn on_connection_established(&mut self, connection: ConnectionRecord) -> bool {
+        match self {
+            // Transform the dial record into a secondary connection.
+            Self::Connected {
+                record,
+                secondary: Some(SecondaryOrDialing::Dialing(dial_record)),
+            } =>
+                if dial_record.connection_id == connection.connection_id {
+                    *self = Self::Connected {
+                        record: record.clone(),
+                        secondary: Some(SecondaryOrDialing::Secondary(connection)),
+                    };
+
+                    return true;
+                },
+            // There's place for a secondary connection.
+            Self::Connected {
+                record,
+                secondary: None,
+            } => {
+                *self = Self::Connected {
+                    record: record.clone(),
+                    secondary: Some(SecondaryOrDialing::Secondary(connection)),
+                };
+
+                return true;
+            }
+
+            // Convert the dial record into a primary connection or preserve it.
+            Self::Dialing { dial_record }
+            | Self::Disconnected {
+                dial_record: Some(dial_record),
+            } =>
+                if dial_record.connection_id == connection.connection_id {
+                    *self = Self::Connected {
+                        record: connection.clone(),
+                        secondary: None,
+                    };
+                    return true;
+                } else {
+                    *self = Self::Connected {
+                        record: connection,
+                        secondary: Some(SecondaryOrDialing::Dialing(dial_record.clone())),
+                    };
+                    return true;
+                },
+
+            Self::Disconnected { dial_record: None } => {
+                *self = Self::Connected {
+                    record: connection,
+                    secondary: None,
+                };
+
+                return true;
+            }
+
+            // Accept the incoming connection.
+            Self::Opening { .. } => {
```

Comment:
I don't think we can "jump" from `Opening` directly to `Connected` following the transport manager logic. May be add a warning?

Reply:
We can still get from `Opening` to `Connected` in the following scenario:
- we are dialing on multiple addresses the remote (opening state outbound)
- the remote establishes a connection with us (inbound)

For this situation, we then cancel the opening futures from the transport manager.

Have added a log trace here and some more state machine documentation, thanks 🙏